src/main/java/software/amazon/event/kafkaconnector/EventBridgeWriter.java

```diff
@@ -127,67 +129,89 @@ public EventBridgeWriter(EventBridgeAsyncClient ebClient, EventBridgeSinkConfig
    * @return list of all records with additional status information
    */
   public List<EventBridgeWriterResult> putItems(List<SinkRecord> records) {
+    return doWithPutEventsRequestFrom(
+        records,
+        (xs, request) -> {
+          try {
+
+            log.trace("EventBridgeWriter sending request to eventbridge: {}", request);
+            var response = ebClient.putEvents(request).get(SDK_TIMEOUT, MILLISECONDS);
+            log.trace("EventBridgeWriter putEvents response: {}", response.entries());
+
+            if (response.failedEntryCount() > 0) {
+              log.warn("Received failed eventbridge entries: {}", response.failedEntryCount());
+            }
+
+            return mapResponseToResult(xs, response);
+          } catch (AwsServiceException
+              | SdkClientException
+              | ExecutionException
+              | InterruptedException
+              | TimeoutException e) {
+
+            var cause = e.getCause();
+            if (cause instanceof EventBridgeException) {
+              var code = ((EventBridgeException) cause).statusCode();
+              // entries size limit exceeded
+              if (code == HttpStatusCode.REQUEST_TOO_LONG) {
+                log.warn("Caught eventbridge exception: batch size limit exceeded", cause);
+                return xs.stream()
+                    .map(
+                        it ->
+                            failure(
+                                it.getSinkRecord(), "413", "EventBridge batch size limit exceeded"))
+                    .collect(toList());
+              }
+            }
+
+            // all other cases assume transient EventBridge service or network error
+            throw new EventBridgeWriterException(e, RETRYABLE);
```

Comment:
@maschnetwork we can eliminate throwing the exception if we include the result and type (retryable/non retryable) to `EventBridgeWriterResult`

Reply:
@agebhar1 `EventBridgeWriterResult` is tied to a specific `SinkRecord` while this exception affects multiple records / batches. Because of this I would rather stick with the exception here. Wdyt?

Reply:
Yes. Currently it is bound to a single `SinkRecord`. Mixing functional style with exceptions isn't a good idea. In case of an exception the data flow is interrupted and jumped to somewhere.

Handle errors in a functional style is using a dedicated datatype like `Either` in Haskell or `Result` in Kotlin. We already introduced such an datatype. Let's extend the datatype by different failure (error) cases. One with with a single `SinkRecord` for JSON serialization, one with multiple `SinkRecord`s for `EventBridgeAsyncClient#putEvents`. The error can also indicate if its fixable or not.

Reply:
Thinking of the upcoming batching logic I agree that the newly created Error class would make it more concise. But to keep the whole new result handling consistent I would rather go for marking each item in the affected batch with the error code instead of creating a new variant in Result that can deal with a list. Basically the same as we do in the case of REQUEST_TOO_LONG.